include/tc/core/cuda/cuda_execution_engine.h

```diff
@@ -31,64 +32,54 @@ namespace tc {
 /// able to execute the kernels for multiple TC i.e. given the language which
 /// can have multiple TCs, people should be able to run things by just calling
 /// out the run function with the name of function and the inputs to run on.
-class CudaExecutionEngine {
+class CudaExecutionEngine : public ExecutionEngine {
  public:
-  struct ExecutorInfo {
-    ExecutorInfo(
+  struct CudaExecutorInfo : public ExecutionEngine::ExecutorInfo {
+    CudaExecutorInfo(
         std::string id,
         std::vector<const DLTensor*> inputsInfo,
-        std::unique_ptr<MappingOptions> options,
+        const MappingOptions& options,
         lang::TreeRef tc,
         size_t handle)
-        : identifier(id),
-          inputsInfo(dlutils::makeDLTensorVector(inputsInfo)),
-          options(std::move(options)),
-          exec(tc, inputsInfo),
-          objectLocalHandle(handle) {}
+        : ExecutionEngine::ExecutorInfo(
+              id,
+              inputsInfo,
+              options.toProtobufSerializedString(),
+              tc,
+              handle) {
+      exec =
+          std::unique_ptr<CudaTcExecutor>(new CudaTcExecutor(tc, inputsInfo));
+    }
 
     void clear() {
-      exec.clearRTC();
+      static_cast<CudaTcExecutor&>(*exec).clearRTC();
     }
-
-    std::string identifier;
-    std::vector<dlutils::DLTensorUPtr> inputsInfo;
-    std::unique_ptr<MappingOptions> options;
-    CudaTcExecutor exec;
-    /// When run is called this is used to find the most recently compiled
-    /// version.
-    size_t objectLocalHandle;
   };
 
   CudaExecutionEngine() = default;
 
-  /// Create the CudaExecutionEngine::tcNameMap_ using the language passed
-  /// to it - should support many TC.
-  void define(const std::string& language);
-
-  /// Create the CudaExecutionEngine::tcNameMap_ from the parsed TC
-  /// string - supports many TC.
-  void define(const std::vector<lang::TreeRef>& treeRefs);
-
   void addTC(const std::string& tc);
 
-  /// Get the output Tensor info that can be used by the calling framework to
-  /// allocate storage for the output.
-  std::vector<const DLTensor*> inferOutputTensorInfo(
-      const std::string& name,
-      const std::vector<const DLTensor*>& inTensorPtrs);
-
   lang::TreeRef treeForFunction(const std::string& name) {
     return tcNameMap_.at(name);
   }
 
   // TODO: Pass autotuning info (none by default, otherwise some struct with
   //       maxtime and other things)
 
-  /// Returns a handle for the compiled kernel
+  /// returns a handle for the compiled kernel
+  /// @{
+  size_t compile(
+      const std::string& name,
+      const std::vector<const DLTensor*>& inputs,
+      const std::string& mappingOptions) override {
+    return compile(name, inputs, MappingOptions(mappingOptions));
+  }
   size_t compile(
       const std::string& name,
       const std::vector<const DLTensor*>& inputs,
       const MappingOptions& options);
+  /// @}
```

Comment:
Why do you need two versions of this method to be able
to extract the abstract parts?
Same for `run`.